### PR TITLE
[DSCP-396] Remove inner nodes from Merkle proofs

### PR DESCRIPTION
**Pull request by flyingleafe**
### Description

1) Merkle proofs do not have inned Merkle signatures anymore
2) Some functions working with Merkle proofs are optimized
3) Refactored the code accordingly to the style guide
4) Moved `Serialise` instances for Merkle datatypes to `Dscp.Crypto.Serialise`
5) and other changes which followed

### YT issue

https://issues.serokell.io/issue/DSCP-396

### Checklist

Hint: a perfect PR has all the checkmarks set.

Possible related changes (conditional):
- [x] I added tests if required, in case they are:
  - Covering new introduced functionality (feature).
  - Regression tests preventing the bug from silently reappearing again (bugfix).
- [x] I have checked the [sample config](../tree/master/docs/config-full-sample.yaml) and [launch scripts](../tree/master/scripts/launch) and updated them if it was necessary (especially if executables or CLIs were changed).
- [x] I have checked READMEs and changed them accordingly if it's necessary (the main [README.md](../tree/master/README.md) as well as subproject READMEs for all related executables).
- [x] If a public API structure or/and data serialization was changed, I made sure that these changes are reflected in the:
  - [Swagger](../tree/master/specs/disciplina) API specs.
  - Any [related documentation](../tree/master/docs/api-types.md).
- [x] If any public documentation was written, I have spellchecked it.

Stylistic (obligatory):
- [x] My commit history is clean, descriptive and do not contain merge or revert commits or commits like "WIP".
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).
- [x] My code is documented (haddock) and these docs are decent (full enough and spellchecked).

### CODE REVIEW

File: witness/tests/Test/Dscp/Witness/Explorer/Explorer.hs [hunk 1]
```diff
@@ -92,13 +92,19 @@ data SpoilVariant
...
-    :: MerkleProof a
+    :: (Eq a, Arbitrary a)
+    => MerkleProof a
     -> Gen (MerkleProof a)
-spoilMerkleProof mproof = do
-    newSig <- arbitrary
-    pure $ mproof { pnSig = newSig }
+spoilMerkleProof (ProofBranch l r) = do
+    var <- choose (0 :: Int, 1)
+    if var == 0
```
Comment (Martoon-00) [inline witness/tests/Test/Dscp/Witness/Explorer/Explorer.hs]: Why not `var :: Bool`? (if you find this more aesthetically correct, I don't object though)

File: core/src/Dscp/Crypto/MerkleTree.hs [hunk 2]
```diff
@@ -85,80 +84,106 @@ instance Foldable MerkleTree where
...
+-- | Construct a @'MerkleTree'@ over a @'Container'@
 fromContainer :: (HasHash (Element t), Container t) => t -> MerkleTree (Element t)
 fromContainer = fromList . toList
 
+-- | Construct a @'MerkleTree'@ over a list.
 fromList :: HasHash a => [a] -> MerkleTree a
-fromList [] = MerkleEmpty
-fromList ls = MerkleTree (nodeFromList ls)
+fromList []       = MerkleEmpty
+fromList (a : as) = MerkleTree $ nodeFromList (a :| as)
```
Comment (Martoon-00) [inline core/src/Dscp/Crypto/MerkleTree.hs]: I'd appreciate if you make it part of `IsList`, I got so many :fire_engine: from this `fromList` at top level.

**Review by Martoon-00 — CHANGES_REQUESTED** [review]
Looks good, just address some minor changes.

File: core/src/Dscp/Crypto/MerkleTree.hs [hunk 3]
```diff
@@ -167,212 +192,260 @@ nodeFromList lst = tree
...
-mkMerkleProofSingle t n = mkMerkleProof t (Set.fromList [n])
-
-mkMerkleProof :: forall a. MerkleTree a -- ^ merkle tree we want to construct a proof from
-                        -> Set LeafIndex -- ^ leaf index used for proof
-                        -> Maybe (MerkleProof a)
+instance Foldable MerkleProof where
+    foldMap _ ProofPruned{} = mempty
+    foldMap f (ProofLeaf v) = f v
+    foldMap f (ProofBranch l r) =
+        F.foldMap f l `mappend` F.foldMap f r
```
[the remark this answers is not in the document]
Comment (flyingleafe) [reply core/src/Dscp/Crypto/MerkleTree.hs]: How so? It's linear, `mappend` for `Endo` is just a function composition, this approach is kind of a standard way of defining `Foldable` for tree-like structures: http://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Foldable.html#t:Foldable

File: core/src/Dscp/Crypto/MerkleTree.hs (hunk 2 above)
Comment (Martoon-00) [reply core/src/Dscp/Crypto/MerkleTree.hs]: Ah right, I missed the fact that any reasonable use of `foldMap` is with `Endo` as its monoid.

File: core/src/Dscp/Crypto/MerkleTree.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Martoon-00) [reply core/src/Dscp/Crypto/MerkleTree.hs]: Ah right, I missed the fact that any reasonable use of `foldMap` is with `Endo` as its monoid.

**Comment by flyingleafe** [comment]
@Martoon-00 bump

**Review by Martoon-00 — APPROVED** [review]
Looks pretty nice to me, thanks.

**Pull request merged**